api/middleware/logger.go

```diff
@@ -24,10 +26,35 @@ type LoggingResponseWriter struct {
 	errBody []byte // the body in case it is an error
 }
 
+func decompressGzip(b []byte) ([]byte, error) {
+	reader, err := gzip.NewReader(bytes.NewReader(b))
+	if err != nil {
+		return nil, err
+	}
+
+	return ioutil.ReadAll(reader)
+}
+
 func (rw *LoggingResponseWriter) Write(b []byte) (int, error) {
 	if rw.ResponseWriter.Status() >= 400 {
-		rw.errBody = make([]byte, len(b))
-		copy(rw.errBody, b)
+		if rw.Header().Get("Content-Encoding") == "gzip" {
```

Comment:
This shouldnt be done here. We should just store the compressed payload in rw.errBody and do the decompress when we write the log message.

The client wont get their response until we call the rw.responseWriter(b),  so we should call it as soon as possible.

The log message is not written until after the request has been completely sent to the client, so spending the time to decompress the data their wont impact request latency (yes, failed responses are likely only going to have small payloads and so it probably wont matter where we do the decompress)

Reply:
Fixed in ff4b0e5